utils/geometry/custom_geometry.py

```diff
@@ -5,6 +5,261 @@
 import matplotlib.pyplot as plt
 import os
 
+class GmshGeometry3D(Geometry):
+    def __init__(self, gmsh_model, external_dim_size=None):
+        self.gmsh_model = gmsh_model
+        self.external_dim_size = external_dim_size
+        self.boundary_normal_global = self.fun_boundary_normal_global()
+        self.bbox = (np.array([1,1,1]))
+        super(GmshGeometry3D, self).__init__(
+            3, self.bbox, 1
+        )
+    
+    def inside(self, x):
+        """Check if x is inside the geometry (including the boundary)."""
+
+        node_tag, node_coords_all, parametricCoord  = self.gmsh_model.mesh.getNodes(dim=self.dim, tag=-1, includeBoundary=True)
+        node_tag_inside = self.gmsh_model.mesh.getNodes(dim=self.dim, tag=-1, includeBoundary=False)[0]
+
+        _, _, node_coords_xyz_inside = self.order_coordinates(node_coords_all, node_tag, node_tag_inside=node_tag_inside)
+
+        if self.external_dim_size:
+            node_coords_xyz_inside = self.add_external_dim(node_coords_xyz_inside)
+
+        return self.is_in_tolerance(x, node_coords_xyz_inside)
+
+    def on_boundary(self, x):
+        """Check if x is on the geometry boundary."""
+
+        node_tag, node_coords_all, _  = self.gmsh_model.mesh.getNodes(dim=self.dim, tag=-1, includeBoundary=True)
+        node_tag_inside = self.gmsh_model.mesh.getNodes(dim=self.dim, tag=-1, includeBoundary=False)[0]
+        node_tag_boundary = np.setdiff1d(node_tag, node_tag_inside)
+
+        _, node_coords_xyz_boundary, _ = self.order_coordinates(node_coords_all, node_tag, node_tag_boundary, node_tag_inside)
+
+        if self.external_dim_size:
+            node_coords_xyz_boundary = self.add_external_dim(node_coords_xyz_boundary)
+        
+        return self.is_in_tolerance(x, node_coords_xyz_boundary)
+    
+    def is_in_tolerance(self, provided, target):
+            
+        tolerance = 1e-5
+        
+        contained_rows = np.all(np.isclose(provided[:, None, :], target, rtol=tolerance, atol=tolerance), axis=2)
+        contained_indices_boolean = np.any(contained_rows, axis=1)
+        return contained_indices_boolean
+        
+    def boundary_normal(self, x):
+        """Slice the unit normal at x for Neumann or Robin boundary conditions."""
+
+        n, _, _, uniq = self.boundary_normal_global
```

Comment:
If I see it correctly, the calls to `self.boundary_normal_global` are independent of x, as you always compute the boundary normals and tangential on the whole geometry, which means that the output of this function will always be the same, no matter when it's called unless the underlying gmsh geometry changes.

Yet, you decided to call it here in `boundary_normal()` as well as `boundary_tangential_1()` and `boundary_tangential_2()` which -- if I understand correctly -- are queried multiple times with different input points during training. 

So from a performance point of view, I was wondering if it might make sense to call `self.fun_boundary_normal_global` once after initialization of the gmsh model and store the results internally and then only perform the slicing operations in `boundary_normal()`, `boundary_tangential_1()` and `boundary_tangential_2()`.